### PR TITLE
[don't merge yet] Improve text generation of DamageTargetEffect

### PR DESCRIPTION
**Pull request by Susucre**
Added support for various distinct wording of DamageTargetEffect, when a DynamicValue is involved.

Fixed rule generation for a bunch of cards using that effect.

### CODE REVIEW

**Comment by Susucre** [comment]
Example of cards properly supported:
[[Lotleth Giant]] [[Noise Marine]] [[Ossuary Rats]]

File: Mage/src/main/java/mage/abilities/dynamicvalue/common/CountersSourceCount.java [hunk 1]
```diff
@@ -47,8 +51,20 @@ public String toString() {
         return "1";
     }
 
+    public CountersSourceCount setSourceName(String sourceName) {
+        this.sourceName = sourceName;
+        return this;
+    }
+
+    public CountersSourceCount doPluralizeCounter() {
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/dynamicvalue/common/CountersSourceCount.java:59]: Try to use `withXXX` for all optional fields with self/this return.

File: Mage/src/main/java/mage/abilities/dynamicvalue/common/CountersSourceCount.java [hunk 2]
```diff
@@ -11,13 +11,17 @@
 public class CountersSourceCount implements DynamicValue {
 
     private final CounterType counterType;
+    private String sourceName = "{this}"; // for text generation
+    private boolean pluralizeCounter = false; // for text generation
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/dynamicvalue/common/CountersSourceCount.java:15]: string counterName = "counter";
withCounterName("counters")

File: Mage/src/main/java/mage/abilities/dynamicvalue/common/SourcePermanentPowerCount.java [hunk 3]
```diff
@@ -13,6 +13,7 @@
 public class SourcePermanentPowerCount implements DynamicValue {
 
     private final boolean allowNegativeValues;
+    private String sourcePossessiveForm = "{this}'s"; // for text generation
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/dynamicvalue/common/SourcePermanentPowerCount.java:16]: sourceName = "{this}'s"
withSourceName("its")

File: Mage/src/main/java/mage/abilities/effects/common/DamageTargetEffect.java [hunk 4]
```diff
@@ -152,13 +173,23 @@ public String getText(Mode mode) {
...
         String message = amount.getMessage();
         sb.append(this.sourceName).append(" deals ");
-        if (message.isEmpty() || !message.equals("1")) {
+        if(whereXWording) {
+            sb.append("X ");
+        }
+        else if(OneDamageForEachWording) {
+            sb.append("1 ");
+        }
+        else if ((amount instanceof StaticValue) && !message.equals("1")) {
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/effects/common/DamageTargetEffect.java:182]: Too much params for same thing (else/else/else). Try to use one param or add runtime check for correct usage (your current text gen un-support multiple usage of that params). One param example:
withAmountText("X")
withAmountText("1")
withAmountText("") // use amount value

Also I recommends to add examples for each text gen param like "param def; // 1 damage -> x damage".

`DamageTargetEffect` is good example of overwhelming of text gen params, it's very hard to use now, even after look at text gen code.

**Review by JayDi85 — CHANGES_REQUESTED** [review]

File: Mage/src/main/java/mage/abilities/effects/common/DamageTargetEffect.java (hunk 4 above)
Comment (JayDi85) [reply Mage/src/main/java/mage/abilities/effects/common/DamageTargetEffect.java:182]: Maybe @xenohedron can help with the ideas here

Comment (xenohedron) [reply Mage/src/main/java/mage/abilities/effects/common/DamageTargetEffect.java:182]: Yeah I'll be back tomorrow and will take a look at these

**Comment by xenohedron** [comment]
I think DamageTargetEffect needs some cleanup before even getting into the text generation.

Also, I think text generation for DynamicValues needs discussion in the general case, not just in this effect.

Will review further later.

**Pull request closed without merging**